### PR TITLE
Fix potential unaligned access in challenge_auth.

### PR DESCRIPTION
**Pull request by jyao1**
Fix: https://github.com/DMTF/libspdm/issues/1160

Add libspdm_{read|write}_uint{16|32|64}

It can be used for unaligned access or big endian later.

Signed-off-by: Jiewen Yao <jiewen.yao@intel.com>

### CODE REVIEW

File: include/library/spdm_common_lib.h [hunk 1]
```diff
@@ -1132,4 +1132,61 @@ uint32_t libspdm_read_uint24(const uint8_t *buffer);
...
+ **/
+uint16_t libspdm_read_uint16(const uint8_t *buffer);
+
+/**
+ * Writes a 16-bit value to memory that may be unaligned.
+ *
+ * @param  buffer  The pointer to a 16-bit value that may be unaligned.
+ * @param  value   16-bit value to write to buffer.
+ *
+ * @return The 16-bit value to write to buffer.
```
Comment (steven-bellock) [inline include/library/spdm_common_lib.h:1150]: Remove line.

File: include/library/spdm_common_lib.h [hunk 2]
```diff
@@ -1132,4 +1132,61 @@ uint32_t libspdm_read_uint24(const uint8_t *buffer);
...
+ **/
+uint32_t libspdm_read_uint32(const uint8_t *buffer);
+
+/**
+ * Writes a 32-bit value to memory that may be unaligned.
+ *
+ * @param  buffer  The pointer to a 32-bit value that may be unaligned.
+ * @param  value   32-bit value to write to buffer.
+ *
+ * @return The 32-bit value to write to buffer.
```
Comment (steven-bellock) [inline include/library/spdm_common_lib.h:1169]: Remove line.

File: include/library/spdm_common_lib.h [hunk 3]
```diff
@@ -1132,4 +1132,61 @@ uint32_t libspdm_read_uint24(const uint8_t *buffer);
...
+ **/
+uint64_t libspdm_read_uint64(const uint8_t *buffer);
+
+/**
+ * Writes a 64-bit value to memory that may be unaligned.
+ *
+ * @param  buffer  The pointer to a 64-bit value that may be unaligned.
+ * @param  value   64-bit value to write to buffer.
+ *
+ * @return The 64-bit value to write to buffer.
```
Comment (steven-bellock) [inline include/library/spdm_common_lib.h:1188]: Remove line.

File: library/spdm_common_lib/libspdm_com_support.c [hunk 4]
```diff
@@ -89,6 +89,102 @@ void libspdm_write_uint24(uint8_t *buffer, uint32_t value)
...
+    return (uint16_t)(buffer[0] | buffer[1] << 8);
+}
+
+/**
+ * Writes a 16-bit value to memory that may be unaligned.
+ *
+ * @param  buffer  The pointer to a 16-bit value that may be unaligned.
+ * @param  value   16-bit value to write to buffer.
+ *
+ * @return The 16-bit value to write to buffer.
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_support.c:110]: Remove line.

File: library/spdm_common_lib/libspdm_com_support.c [hunk 5]
```diff
@@ -89,6 +89,102 @@ void libspdm_write_uint24(uint8_t *buffer, uint32_t value)
...
+ **/
+void libspdm_write_uint16(uint8_t *buffer, uint16_t value)
+{
+    buffer[0] = (uint8_t)(value & 0xFF);
+    buffer[1] = (uint8_t)((value >> 8) & 0xFF);
+    return;
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_support.c:116]: Remove unneeded `return`.

File: library/spdm_common_lib/libspdm_com_support.c [hunk 6]
```diff
@@ -89,6 +89,102 @@ void libspdm_write_uint24(uint8_t *buffer, uint32_t value)
...
+    return (uint32_t)(buffer[0] | buffer[1] << 8 | buffer[2] << 16 | buffer[3] << 24);
+}
+
+/**
+ * Writes a 32-bit value to memory that may be unaligned.
+ *
+ * @param  buffer  The pointer to a 32-bit value that may be unaligned.
+ * @param  value   32-bit value to write to buffer.
+ *
+ * @return The 32-bit value to write to buffer.
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_support.c:137]: Remove line.

File: library/spdm_common_lib/libspdm_com_support.c [hunk 7]
```diff
@@ -89,6 +89,102 @@ void libspdm_write_uint24(uint8_t *buffer, uint32_t value)
...
+ **/
+void libspdm_write_uint32(uint8_t *buffer, uint32_t value)
+{
+    buffer[0] = (uint8_t)(value & 0xFF);
+    buffer[1] = (uint8_t)((value >> 8) & 0xFF);
+    buffer[2] = (uint8_t)((value >> 16) & 0xFF);
+    buffer[3] = (uint8_t)((value >> 24) & 0xFF);
+    return;
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_support.c:145]: Remove unneeded return.

File: library/spdm_common_lib/libspdm_com_support.c [hunk 8]
```diff
@@ -89,6 +89,102 @@ void libspdm_write_uint24(uint8_t *buffer, uint32_t value)
...
+           ((uint64_t)(buffer[7]) << 56);
+}
+
+/**
+ * Writes a 64-bit value to memory that may be unaligned.
+ *
+ * @param  buffer  The pointer to a 64-bit value that may be unaligned.
+ * @param  value   64-bit value to write to buffer.
+ *
+ * @return The 64-bit value to write to buffer.
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_support.c:173]: Remove line.

File: library/spdm_common_lib/libspdm_com_support.c [hunk 9]
```diff
@@ -89,6 +89,102 @@ void libspdm_write_uint24(uint8_t *buffer, uint32_t value)
...
+{
+    buffer[0] = (uint8_t)(value & 0xFF);
+    buffer[1] = (uint8_t)((value >> 8) & 0xFF);
+    buffer[2] = (uint8_t)((value >> 16) & 0xFF);
+    buffer[3] = (uint8_t)((value >> 24) & 0xFF);
+    buffer[4] = (uint8_t)((value >> 32) & 0xFF);
+    buffer[5] = (uint8_t)((value >> 40) & 0xFF);
+    buffer[6] = (uint8_t)((value >> 48) & 0xFF);
+    buffer[7] = (uint8_t)((value >> 56) & 0xFF);
+    return;
```
Comment (steven-bellock) [inline library/spdm_common_lib/libspdm_com_support.c:185]: Remove unneeded return.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

**Comment by steven-bellock** [comment]
The pattern `*(uintXX_t *)` is used sparsely enough that this should be ok in the near-term. However we should migrate to something more optimized. In particular if a system is both little-endian and allows for unaligned accesses then there should be no additional overhead from using these functions. This would be accomplished by 1. macro definitions and 2. placing the functions in headers and have them made `inline`. (If you believe in that keyword.)

File: include/library/spdm_common_lib.h (hunk 1 above)
Comment (jyao1) [reply include/library/spdm_common_lib.h]: done

File: include/library/spdm_common_lib.h (hunk 2 above)
Comment (jyao1) [reply include/library/spdm_common_lib.h]: done

File: include/library/spdm_common_lib.h (hunk 3 above)
Comment (jyao1) [reply include/library/spdm_common_lib.h]: done

File: library/spdm_common_lib/libspdm_com_support.c (hunk 4 above)
Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_support.c]: done

File: library/spdm_common_lib/libspdm_com_support.c (hunk 5 above)
Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_support.c]: done

File: library/spdm_common_lib/libspdm_com_support.c (hunk 7 above)
Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_support.c]: done

File: library/spdm_common_lib/libspdm_com_support.c (hunk 8 above)
Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_support.c]: done

File: library/spdm_common_lib/libspdm_com_support.c (hunk 9 above)
Comment (jyao1) [reply library/spdm_common_lib/libspdm_com_support.c]: done

**Review by steven-bellock — APPROVED** [review]

**Pull request merged**